lambdas/data-migration2/src/granulesAndFiles.ts

```diff
@@ -0,0 +1,208 @@
+    { name, version }
+  );
+
+  const existingRecord = await knex<PostgresGranuleRecord>('granules')
+    .where({ granule_id: record.granuleId, collection_cumulus_id: collectionCumulusId })
+    .first();
+
+  // Throw error if it was already migrated.
+  if (existingRecord) {
+    throw new RecordAlreadyMigrated(`Granule ${record.granuleId} was already migrated, skipping`);
+  }
+
+  const granule = await translateApiGranuleToPostgresGranule(record, knex, collectionCumulusId);
+  await knex(tableNames.granules).insert(granule);
+};
+
+/**
+ * Migrate File record from a Granules record from DynamoDB  to RDS.
+ *
+ * @param {ApiFile} file - Granule file
+ * @param {string} granuleId - ID of granule
+ * @param {string} collectionId - ID of collection
+ * @param {Knex} knex - Knex client for writing to RDS database
+ * @returns {Promise<number>} - Cumulus ID for record
+ * @throws {RecordAlreadyMigrated} if record was already migrated
+ */
+export const migrateFileRecord = async (
+  file: ApiFile,
+  granuleId: string,
+  collectionId: string,
+  knex: Knex
+): Promise<void> => {
+  const [name, version] = collectionId.split('___');
+  const collectionPgModel = new CollectionPgModel();
+  const granulePgModel = new GranulePgModel();
+
+  const collectionCumulusId = await collectionPgModel.getRecordCumulusId(
+    knex,
+    { name, version }
+  );
+
+  const granuleCumulusId = await granulePgModel.getRecordCumulusId(
+    knex,
+    { granule_id: granuleId, collection_cumulus_id: collectionCumulusId }
+  );
+
+  const bucket = getBucket(file);
+  const key = getKey(file);
+
+  // Map old record to new schema.
+  const updatedRecord: PostgresFile = {
+    bucket,
+    key,
+    granule_cumulus_id: granuleCumulusId,
+    file_size: file.size,
+    checksum_value: file.checksum,
+    checksum_type: file.checksumType,
+    file_name: file.fileName,
+    source: file.source,
+    path: file.path,
+  };
+  await knex(tableNames.files).insert(updatedRecord);
+};
+
+/**
+ * Migrate granule and files from DynamoDB to RDS
+ * @param {AWS.DynamoDB.DocumentClient.AttributeMap} dynamoRecord
+ * @param {Knex} knex
+ * @param {GranulesAndFilesMigrationSummary} granuleAndFileMigrationSummary
+ * @returns {Promise<MigrationSummary>} - Migration summary for files
+ */
+export const migrateGranuleAndFilesViaTransaction = async (
+  dynamoRecord: AWS.DynamoDB.DocumentClient.AttributeMap,
+  knex: Knex,
+  granuleAndFileMigrationSummary: GranulesAndFilesMigrationSummary
+): Promise<GranulesAndFilesMigrationSummary> => {
+  const files = dynamoRecord.files;
+  const granuleId = dynamoRecord.granuleId;
+  const collectionId = dynamoRecord.collectionId;
+  const { granulesSummary, filesSummary } = granuleAndFileMigrationSummary;
+
+  try {
+    granulesSummary.dynamoRecords += 1;
+    await migrateGranuleRecord(dynamoRecord, knex);
+    granulesSummary.success += 1;
+    await Promise.all(files.map(async (file : ApiFile) => {
+      filesSummary.dynamoRecords += 1;
```

Comment:
So we really need to use a `knex.transaction` here. A try/catch doesn't quite achieve the same thing. With the try/catch as written, if `migrateGranuleRecord` succeeds but then `migrateFileRecord` errors, the granule record that was written by `migrateGranuleRecord` still exists. Thus we have "partial data" where the granule record was written but the files were not, which we previously decided that we do not want.

When you use `knex.transaction` to surround these writes, then if `migrateGranuleRecord` succeeds but `migrateFileRecord` (for any file) fails, then the write for the granule is **not committed**, thus both the granule and files are not written.

Does that make sense?

Reply:
Also, if we use a transaction, then the success/failure of writes is "all or nothing". Either the granule and all of the file records are written or none of them are. Thus, I don't think you would need to do the `+= 1` operation to count the number of records that succeeded or failed. If the transaction didn't thrown an error, then the granule was a success and all of its files were, otherwise if an error was thrown than they were all failures. It should simplify the logic here a bit

Reply:
Kind of like this? https://github.com/nasa/cumulus/blob/b6542f89bd4d2a5b8be9805930468952b2659c6c/lambdas/data-migration2/src/granulesAndFiles.ts#L137-L154 @markdboyd

Reply:
In this case it was easier to just create a PR to show what I meant:

https://github.com/nasa/cumulus/pull/2056

This PR doesn't take into consideration addressing other feedback that I left but just shows how you could use the transaction